plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/RewriteLikeAndEscapeWithCaseSensitivity.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.jdbc.expression;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.matching.Capture;
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.plugin.base.expression.ConnectorExpressionRule;
+import io.trino.plugin.jdbc.CaseSensitivity;
+import io.trino.plugin.jdbc.JdbcColumnHandle;
+import io.trino.plugin.jdbc.QueryParameter;
+import io.trino.spi.expression.Call;
+import io.trino.spi.expression.ConnectorExpression;
+import io.trino.spi.expression.Variable;
+import io.trino.spi.type.VarcharType;
+
+import java.util.Optional;
+
+import static io.trino.matching.Capture.newCapture;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.argument;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.argumentCount;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.call;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.expression;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.functionName;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.type;
+import static io.trino.spi.expression.StandardFunctions.LIKE_FUNCTION_NAME;
+import static io.trino.spi.type.BooleanType.BOOLEAN;
+import static java.lang.String.format;
+
+public class RewriteLikeAndEscapeWithCaseSensitivity
+        implements ConnectorExpressionRule<Call, ParameterizedExpression>
+{
+    private static final Capture<ConnectorExpression> LIKE_VALUE = newCapture();
+    private static final Capture<ConnectorExpression> LIKE_PATTERN = newCapture();
+    private static final Pattern<Call> PATTERN = call()
+            .with(functionName().equalTo(LIKE_FUNCTION_NAME))
+            .with(type().equalTo(BOOLEAN))
+            .with(argumentCount().matching(count -> count == 2 || count == 3))
+            .with(argument(0).matching(expression().capturedAs(LIKE_VALUE).with(type().matching(VarcharType.class::isInstance))))
+            .with(argument(1).matching(expression().capturedAs(LIKE_PATTERN).with(type().matching(VarcharType.class::isInstance))));
+    // TODO add capture on ESCAPE argument and type
```

Comment:
We capture first 2 arguments, and third optional we took from arguments.

I cannot add third capture, because in this case - pattern matching will not be able to capture call with 2 arguments.

I know that it's optional - 
```
   public static Property<Call, ?, ConnectorExpression> argument(int argument)
    {
        checkArgument(0 <= argument, "Invalid argument index: %s", argument);
        return Property.optionalProperty(format("argument(%s)", argument), call -> {
            if (argument < call.getArguments().size()) {
                return Optional.of(call.getArguments().get(argument));
            }
            return Optional.empty();
        });
    }
```

But it didn't work, maybe some bug in parsing.
I am investigating this and will create issue

Reply:
Ok I think I need to remove todo, because it's not a bug, it's just the way how things works:
Inside `WithPattern`
```

@Override
    public <C> Stream<Match> accept(Object object, Captures captures, C context)
    {
        //TODO remove cast
        BiFunction<? super T, C, Optional<?>> property = (BiFunction<? super T, C, Optional<?>>) propertyPattern.getProperty().getFunction();
        Optional<?> propertyValue = property.apply((T) object, context);
        return propertyValue.map(value -> propertyPattern.getPattern().match(value, captures, context))
                .orElseGet(Stream::of);
    }
```

So if `propertyValue` is empty or `propertyPattern.getPattern().match(value, captures, context)` - it's the same, we didn't capture anything and returning empty stream and stop the flow.

Reply:
So there is no such thing like "optional" capture argument - if you add it it will be captured or not.

Reply:
This PR could be done as follow up - https://github.com/trinodb/trino/pull/18481

Reply:
The simpler solution would be to just split into two rules - one for LIKE and one for LIKE with ESCAPE. Like how we have in the existing DSL based rules.

Reply:
Changing `Pattern` so that it can accept partial matches without explicitly asking for partial matches looks risky.

Reply:
But it would result in some redundant code right ?

Reply:
For this PR however consider this comment resolved since the result is that the TODO is applicable.

Reply:
@hashhar thx! refactored to separate rules
cc @Praveen2112